pkg/gardenlet/controller/federatedseed/seed_control.go

```diff
@@ -0,0 +1,238 @@
+
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/shootstate"
+	"github.com/gardener/gardener/pkg/logger"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller is responsible for maintaining multiple federated Seeds' controllers
+type Controller struct {
+	federatedSeedControllerManager *federatedSeedControllerManager
+	k8sGardenClient                kubernetes.Interface
+
+	config *config.GardenletConfiguration
+
+	seedLister    gardencorelisters.SeedLister
+	seedHasSynced cache.InformerSynced
+	seedQueue     workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+
+	ctx                    context.Context
+	numberOfRunningWorkers int
+	workerCh               chan int
+}
+
+// NewFederatedSeedController creates new controller that syncs extensions states to the ShootState resource in the garden cluster.
+func NewFederatedSeedController(ctx context.Context, k8sGardenClient kubernetes.Interface, gardenCoreInformerFactory gardencoreinformers.SharedInformerFactory, config *config.GardenletConfiguration, recorder record.EventRecorder) *Controller {
+	var seedInformer = gardenCoreInformerFactory.Core().V1alpha1().Seeds()
+
+	controller := &Controller{
+		k8sGardenClient: k8sGardenClient,
+		seedLister:      seedInformer.Lister(),
+		seedQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "seed-controllers"),
+		config:          config,
+		ctx:             ctx,
+		workerCh:        make(chan int),
+		recorder:        recorder,
+	}
+
+	controller.federatedSeedControllerManager = &federatedSeedControllerManager{
+		seedShootStateSyncController: make(map[string]*shootstate.SyncController),
+	}
+
+	seedInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controllerutils.SeedFilterFunc(confighelper.SeedNameFromSeedConfig(config.SeedConfig), config.SeedSelector),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.seedAdd,
+			UpdateFunc: controller.seedUpdate,
+			DeleteFunc: controller.seedDelete,
+		},
+	})
+
+	controller.seedHasSynced = seedInformer.Informer().HasSynced
+	return controller
+}
+
+func (c *Controller) seedAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+
+		return
+	}
+	c.seedQueue.Add(key)
+}
+
+func (c *Controller) seedUpdate(oldObj, newObj interface{}) {
+	var (
+		newSeedObj = newObj.(*gardencorev1alpha1.Seed)
+		oldSeedObj = oldObj.(*gardencorev1alpha1.Seed)
+	)
+
+	newSeedBootstrappedCondition := helper.GetCondition(newSeedObj.Status.Conditions, gardencorev1alpha1.SeedBootstrapped)
```

Comment:
GetCondition can return nil, imho we should add nil checks in case the bootstrapped condition has not been initialised.